src/utils/simplifier.js

```diff
@@ -104,14 +104,50 @@ function simplify(realm, value: Value, isCondition: boolean = false): Value {
         // (x: boolean) || false <=> x
         if (!y.mightNotBeFalse()) return op === "||" ? x : realm.intrinsics.false;
       }
-      if (
-        op === "||" &&
-        y instanceof AbstractValue &&
-        y.kind === "||" &&
-        x.equals(y.args[0]) &&
-        !y.args[1].mightNotBeTrue()
-      )
-        return y;
+      if (op === "||") {
+        if (y instanceof AbstractValue && y.kind === "||") {
+          // x || x || y <=> x || y
+          if (x.equals(y.args[0])) return y;
+          if (x instanceof AbstractValue && x.kind === "!") {
+            // !x0 || y0 || x0 <=> true, if isCondition
+            if (isCondition && x.args[0].equals(y.args[1])) return realm.intrinsics.true;
+          }
+        }
+      }
+      if (op === "&&") {
+        if (x instanceof AbstractValue) {
+          if (x.kind === "&&") {
+            // (x && y) && x <=> x && y
+            // (x && y) && y <=> x && y
+            if (x.args[0].equals(y) || x.args[1].equals(y)) return x;
```

Comment:
Thinking about it, the first one is actually only valid if `isCondition` without reordering: `(1 && 2) && 1` vs `1 && 2`.

Reply:
Good catch.